### ISSUE TITLE
A lot of Socket.ThrowIfDisposedAndClosed() exceptions

### ISSUE THREAD

**Issue by VictorKochetkov** [rabbitmq/rabbitmq-dotnet-client#423]
When using RabbitMQ client in Xamarin.Android application there is a huge amount of exceptions occured randomatically - `Socket.ThrowIfDisposedAndClosed()`.
Stack trace:

> Xamarin caused by: android.runtime.JavaProxyThrowable: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
  at System.Net.Sockets.Socket.ThrowIfDisposedAndClosed () [0x0001b] in <88c7ebccfa8943c6b6cb37a0e23aa47a>:0 
  at System.Net.Sockets.Socket.EndConnect (System.IAsyncResult asyncResult) [0x00000] in <88c7ebccfa8943c6b6cb37a0e23aa47a>:0 
  at System.Net.Sockets.SocketTaskExtensions+<>c.<ConnectAsync>b__3_1 (System.IAsyncResult asyncResult) [0x00006] in <88c7ebccfa8943c6b6cb37a0e23aa47a>:0 
  at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncCoreLogic (System.IAsyncResult iar, System.Func`2[T,TResult] endFunction, System.Action`1[T] endAction, System.Threading.Tasks.Task`1[TResult] promise, System.Boolean requiresSynchronization) [0x00019] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at RabbitMQ.Client.TcpClientAdapter+<ConnectAsync>d__2.MoveNext () [0x0011d] in <6475a35b2e2f49ac8021508f877d5869>:0 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
  at System.Net.Sockets.Socket.ThrowIfDisposedAndClosed () [0x0001b] in <88c7ebccfa8943c6b6cb37a0e23aa47a>:0 
  at System.Net.Sockets.Socket.EndConnect (System.IAsyncResult asyncResult) [0x00000] in <88c7ebccfa8943c6b6cb37a0e23aa47a>:0 
  at System.Net.Sockets.SocketTaskExtensions+<>c.<ConnectAsync>b__3_1 (System.IAsyncResult asyncResult) [0x00006] in <88c7ebccfa8943c6b6cb37a0e23aa47a>:0 
  at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncCoreLogic (System.IAsyncResult iar, System.Func`2[T,TResult] endFunction, System.Action`1[T] endAction, System.Threading.Tasks.Task`1[TResult] promise, System.Boolean requiresSynchronization) [0x00019] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <841652e4f9cb472e934a994cc6cfa0e3>:0 
  at RabbitMQ.Client.TcpClientAdapter+<ConnectAsync>d__2.MoveNext () [0x0011d] in <6475a35b2e2f49ac8021508f877d5869>:0 <---

This exception is throws somewhere in client in another Task, so it can't be catched.
RabbitMQ client version is `5.1.0-rc1`

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team).

We get at least a dozen of questions through various venues every single day, often light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because GitHub is a tool our team uses heavily nearly every day, the signal/noise ratio of issues is something we care about a lot.

Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Thank you.

**Comment by michaelklishin** [comment]
We cannot suggest much with the amount of information provided. Very likely they are simply operations attempted concurrently on a closed connection, or a connection just recently [detected as closed](http://www.rabbitmq.com/heartbeats.html). Depending on how many tasks and connections there are at the moment of connection closure or missed heartbeats event, the number of tasks that hit this will vary.
